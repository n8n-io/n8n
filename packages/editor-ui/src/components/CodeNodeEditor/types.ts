import type { EditorView } from '@codemirror/view';
import type { I18nClass } from '@/plugins/i18n';
import type { Workflow } from 'n8n-workflow';
import type { Node } from 'estree';
import type { CODE_LANGUAGES, CODE_MODES } from './constants';

export type CodeNodeEditorMixin = Vue.VueConstructor<
	Vue & {
		$locale: I18nClass;
		editor: EditorView | null;
		mode: 'runOnceForAllItems' | 'runOnceForEachItem';
		getCurrentWorkflow(): Workflow;
	}
>;

export type RangeNode = Node & { range: [number, number] };

export type CodeLanguage = (typeof CODE_LANGUAGES)[number];
export type CodeMode = (typeof CODE_MODES)[number];
