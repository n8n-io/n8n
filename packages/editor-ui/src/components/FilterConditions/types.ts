import type { BaseTextKey } from '@/plugins/i18n';
import type { FilterOperatorValue } from 'n8n-workflow';

export interface FilterOperator extends FilterOperatorValue {
	name: BaseTextKey;
}

export interface FilterOperatorGroup {
	id: string;
	name: BaseTextKey;
	icon?: string;
	children: FilterOperator[];
}

export type ConditionResult =
	| { status: 'resolve_error' }
	| { status: 'validation_error'; error: string }
	| { status: 'success'; result: boolean };
