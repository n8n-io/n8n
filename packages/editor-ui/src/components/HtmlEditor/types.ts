export type Range = [number, number];

export type Section = {
	kind: 'html' | 'script' | 'style';
	content: string;
	range: Range;
};
