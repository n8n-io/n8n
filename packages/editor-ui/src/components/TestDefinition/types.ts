export interface TestExecution {
	lastRun: string | null;
	errorRate: number | null;
	metrics: Record<string, number>;
}

export interface TestListItem {
	id: string;
	name: string;
	tagName: string;
	testCases: number;
	execution: TestExecution;
}
