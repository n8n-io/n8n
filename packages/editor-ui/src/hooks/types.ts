import type { ITelemetryTrackProperties } from 'n8n-workflow';

export interface TelemetryEventData {
	eventName: string;
	properties?: ITelemetryTrackProperties;
}
