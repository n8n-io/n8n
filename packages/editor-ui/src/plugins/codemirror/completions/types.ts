import type { resolveParameter } from '@/mixins/workflowHelpers';
import type { DocMetadata } from 'n8n-workflow';

export type Resolved = ReturnType<typeof resolveParameter>;

export type ExtensionTypeName = 'number' | 'string' | 'date' | 'array' | 'object';

export type FnToDoc = { [fnName: string]: { doc?: DocMetadata } };

export type FunctionOptionType = 'native-function' | 'extension-function';
export type KeywordOptionType = 'keyword';
export type AutocompleteOptionType = FunctionOptionType | KeywordOptionType;
