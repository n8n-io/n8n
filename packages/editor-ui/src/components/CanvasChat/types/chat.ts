export interface LangChainMessage {
	id: string[];
	kwargs: {
		content: string;
	};
}

export interface MemoryOutput {
	action: string;
	chatHistory?: LangChainMessage[];
}

export interface IChatMessageResponse {
	executionId?: string;
	success: boolean;
	error?: Error;
}

export interface IChatResizeStyles {
	'--panel-height': string;
	'--chat-width': string;
}
