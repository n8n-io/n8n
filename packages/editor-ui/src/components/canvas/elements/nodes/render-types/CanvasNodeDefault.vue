<script lang="ts" setup>
import { computed, inject, useCssModule } from 'vue';
import { useNodeConnections } from '@/composables/useNodeConnections';
import { CanvasNodeKey } from '@/constants';

const node = inject(CanvasNodeKey);

const $style = useCssModule();

const label = computed(() => node?.label.value ?? '');
const inputs = computed(() => node?.data.value.inputs ?? []);
const outputs = computed(() => node?.data.value.outputs ?? []);

const { mainOutputs } = useNodeConnections({
	inputs,
	outputs,
});

const classes = computed(() => {
	return {
		[$style.node]: true,
		[$style.selected]: node?.selected.value,
	};
});

const styles = computed(() => {
	return {
		'--node-main-output-count': mainOutputs.value.length,
	};
});
</script>

<template>
	<div v-if="node" :class="classes" :style="styles" data-test-id="canvas-node-default">
		<slot />
		<div v-if="label" :class="$style.label">{{ label }}</div>
	</div>
</template>

<style lang="scss" module>
.node {
	height: calc(100px + max(0, var(--node-main-output-count, 1) - 4) * 50px);
	width: 100px;
	display: flex;
	align-items: center;
	justify-content: center;
	background: var(--canvas-node--background, var(--color-canvas-node-background));
	border: 2px solid var(--canvas-node--border-color, var(--color-foreground-xdark));
	border-radius: var(--border-radius-large);
}

.label {
	top: 100%;
	position: absolute;
	font-size: var(--font-size-m);
	text-align: center;
	width: 100%;
	min-width: 200px;
	margin-top: var(--spacing-2xs);
}

.selected {
	box-shadow: 0 0 0 4px var(--color-canvas-selected);
}
</style>
