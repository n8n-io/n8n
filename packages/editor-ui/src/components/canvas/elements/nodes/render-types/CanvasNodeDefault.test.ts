import CanvasNodeDefault from '@/components/canvas/elements/nodes/render-types/CanvasNodeDefault.vue';
import { createComponentRenderer } from '@/__tests__/render';
import { NodeConnectionType } from 'n8n-workflow';
import { createCanvasNodeProvide, createCanvasProvide } from '@/__tests__/data';
import { createTestingPinia } from '@pinia/testing';
import { setActivePinia } from 'pinia';
import { CanvasConnectionMode, CanvasNodeRenderType } from '@/types';

const renderComponent = createComponentRenderer(CanvasNodeDefault, {
	global: {
		provide: {
			...createCanvasProvide(),
		},
	},
});

beforeEach(() => {
	const pinia = createTestingPinia();
	setActivePinia(pinia);
});

describe('CanvasNodeDefault', () => {
	it('should render node correctly', () => {
		const { getByTestId } = renderComponent({
			global: {
				provide: {
					...createCanvasNodeProvide(),
				},
			},
		});

		expect(getByTestId('canvas-default-node')).toMatchSnapshot();
	});

	describe('inputs', () => {
		it('should adjust height css variable based on the number of inputs (1 input)', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								inputs: [{ type: NodeConnectionType.Main, index: 0 }],
							},
						}),
					},
				},
			});

			const nodeElement = getByText('Test Node').closest('.node');
			expect(nodeElement).toHaveStyle({ '--canvas-node--main-input-count': '1' }); // height calculation based on the number of inputs
		});

		it('should adjust height css variable based on the number of inputs (multiple inputs)', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								inputs: [
									{ type: NodeConnectionType.Main, index: 0 },
									{ type: NodeConnectionType.Main, index: 0 },
									{ type: NodeConnectionType.Main, index: 0 },
								],
							},
						}),
					},
				},
			});

			const nodeElement = getByText('Test Node').closest('.node');
			expect(nodeElement).toHaveStyle({ '--canvas-node--main-input-count': '3' }); // height calculation based on the number of inputs
		});
	});

	describe('outputs', () => {
		it('should adjust height css variable based on the number of outputs (1 output)', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								outputs: [{ type: NodeConnectionType.Main, index: 0 }],
							},
						}),
					},
				},
			});

			const nodeElement = getByText('Test Node').closest('.node');
			expect(nodeElement).toHaveStyle({ '--canvas-node--main-output-count': '1' }); // height calculation based on the number of outputs
		});

		it('should adjust height css variable based on the number of outputs (multiple outputs)', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								outputs: [
									{ type: NodeConnectionType.Main, index: 0 },
									{ type: NodeConnectionType.Main, index: 0 },
									{ type: NodeConnectionType.Main, index: 0 },
								],
							},
						}),
					},
				},
			});

			const nodeElement = getByText('Test Node').closest('.node');
			expect(nodeElement).toHaveStyle({ '--canvas-node--main-output-count': '3' }); // height calculation based on the number of outputs
		});
	});

	describe('selected', () => {
		it('should apply selected class when node is selected', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({ selected: true }),
					},
				},
			});
			expect(getByText('Test Node').closest('.node')).toHaveClass('selected');
		});

		it('should not apply selected class when node is not selected', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide(),
					},
				},
			});
			expect(getByText('Test Node').closest('.node')).not.toHaveClass('selected');
		});
	});

	describe('disabled', () => {
		it('should apply disabled class when node is disabled', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								disabled: true,
							},
						}),
					},
				},
			});

			expect(getByText('Test Node').closest('.node')).toHaveClass('disabled');
			expect(getByText('(Deactivated)')).toBeVisible();
		});

		it('should not apply disabled class when node is enabled', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide(),
					},
				},
			});
			expect(getByText('Test Node').closest('.node')).not.toHaveClass('disabled');
		});

		it('should render strike-through when node is disabled and has node input and output handles', () => {
			const { container } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								disabled: true,
								inputs: [{ type: NodeConnectionType.Main, index: 0 }],
								outputs: [{ type: NodeConnectionType.Main, index: 0 }],
								connections: {
									[CanvasConnectionMode.Input]: {
										[NodeConnectionType.Main]: [
											[{ node: 'node', type: NodeConnectionType.Main, index: 0 }],
										],
									},
									[CanvasConnectionMode.Output]: {
										[NodeConnectionType.Main]: [
											[{ node: 'node', type: NodeConnectionType.Main, index: 0 }],
										],
									},
								},
							},
						}),
					},
				},
			});

			expect(container.querySelector('.disabledStrikeThrough')).toBeVisible();
		});
	});

	describe('waiting', () => {
		it('should apply waiting class when node is waiting', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({ data: { execution: { running: true, waiting: '123' } } }),
					},
				},
			});
			expect(getByText('Test Node').closest('.node')).toHaveClass('waiting');
		});
	});

	describe('running', () => {
		it('should apply running class when node is running', () => {
			const { getByText } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({ data: { execution: { running: true } } }),
					},
				},
			});
			expect(getByText('Test Node').closest('.node')).toHaveClass('running');
		});
	});

	describe('configurable', () => {
		it('should render configurable node correctly', () => {
			const { getByTestId } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								render: {
									type: CanvasNodeRenderType.Default,
									options: { configurable: true },
								},
							},
						}),
					},
				},
			});

			expect(getByTestId('canvas-configurable-node')).toMatchSnapshot();
		});

		describe('inputs', () => {
			it('should adjust width css variable based on the number of non-main inputs', () => {
				const { getByText } = renderComponent({
					global: {
						provide: {
							...createCanvasNodeProvide({
								data: {
									inputs: [
										{ type: NodeConnectionType.Main, index: 0 },
										{ type: NodeConnectionType.AiTool, index: 0 },
										{ type: NodeConnectionType.AiDocument, index: 0, required: true },
										{ type: NodeConnectionType.AiMemory, index: 0, required: true },
									],
									render: {
										type: CanvasNodeRenderType.Default,
										options: {
											configurable: true,
										},
									},
								},
							}),
						},
					},
				});

				const nodeElement = getByText('Test Node').closest('.node');
				expect(nodeElement).toHaveStyle({ '--configurable-node--input-count': '3' });
			});
		});
	});

	describe('configuration', () => {
		it('should render configuration node correctly', () => {
			const { getByTestId } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								render: {
									type: CanvasNodeRenderType.Default,
									options: { configuration: true },
								},
							},
						}),
					},
				},
			});

			expect(getByTestId('canvas-configuration-node')).toMatchSnapshot();
		});

		it('should render configurable configuration node correctly', () => {
			const { getByTestId } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								render: {
									type: CanvasNodeRenderType.Default,
									options: { configurable: true, configuration: true },
								},
							},
						}),
					},
				},
			});

			expect(getByTestId('canvas-configurable-node')).toMatchSnapshot();
		});
	});

	describe('trigger', () => {
		it('should render trigger node correctly', () => {
			const { getByTestId } = renderComponent({
				global: {
					provide: {
						...createCanvasNodeProvide({
							data: {
								render: {
									type: CanvasNodeRenderType.Default,
									options: { trigger: true },
								},
							},
						}),
					},
				},
			});

			expect(getByTestId('canvas-trigger-node')).toMatchSnapshot();
		});
	});
});
