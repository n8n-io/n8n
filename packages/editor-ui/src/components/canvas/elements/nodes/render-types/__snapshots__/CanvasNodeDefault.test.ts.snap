// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CanvasNodeDefault > configurable > should render configurable node correctly 1`] = `
<div
  class="node configurable"
  data-test-id="canvas-configurable-node"
  style="--configurable-node--input-count: 0; --canvas-node--main-input-count: 0; --canvas-node--main-output-count: 0;"
>
  <!--v-if-->
  
  
  <!--v-if-->
  <!--v-if-->
  <div
    class="description"
  >
    <div
      class="label"
    >
      Test Node 
      <!--v-if-->
    </div>
    <!--v-if-->
    <div
      class="subtitle"
    >
      Test Node Subtitle
    </div>
  </div>
</div>
`;

exports[`CanvasNodeDefault > configuration > should render configurable configuration node correctly 1`] = `
<div
  class="node configurable configuration"
  data-test-id="canvas-configurable-node"
  style="--configurable-node--input-count: 0; --canvas-node--main-input-count: 0; --canvas-node--main-output-count: 0;"
>
  <!--v-if-->
  
  
  <!--v-if-->
  <!--v-if-->
  <div
    class="description"
  >
    <div
      class="label"
    >
      Test Node 
      <!--v-if-->
    </div>
    <!--v-if-->
    <div
      class="subtitle"
    >
      Test Node Subtitle
    </div>
  </div>
</div>
`;

exports[`CanvasNodeDefault > configuration > should render configuration node correctly 1`] = `
<div
  class="node configuration"
  data-test-id="canvas-configuration-node"
  style="--canvas-node--main-input-count: 0; --canvas-node--main-output-count: 0;"
>
  <!--v-if-->
  
  
  <!--v-if-->
  <!--v-if-->
  <div
    class="description"
  >
    <div
      class="label"
    >
      Test Node 
      <!--v-if-->
    </div>
    <!--v-if-->
    <div
      class="subtitle"
    >
      Test Node Subtitle
    </div>
  </div>
</div>
`;

exports[`CanvasNodeDefault > should render node correctly 1`] = `
<div
  class="node"
  data-test-id="canvas-default-node"
  style="--canvas-node--main-input-count: 0; --canvas-node--main-output-count: 0;"
>
  <!--v-if-->
  
  
  <!--v-if-->
  <!--v-if-->
  <div
    class="description"
  >
    <div
      class="label"
    >
      Test Node 
      <!--v-if-->
    </div>
    <!--v-if-->
    <div
      class="subtitle"
    >
      Test Node Subtitle
    </div>
  </div>
</div>
`;

exports[`CanvasNodeDefault > trigger > should render trigger node correctly 1`] = `
<div
  class="node trigger"
  data-test-id="canvas-trigger-node"
  style="--canvas-node--main-input-count: 0; --canvas-node--main-output-count: 0;"
>
  <!--v-if-->
  
  
  <!--v-if-->
  <!--v-if-->
  <div
    class="description"
  >
    <div
      class="label"
    >
      Test Node 
      <!--v-if-->
    </div>
    <!--v-if-->
    <div
      class="subtitle"
    >
      Test Node Subtitle
    </div>
  </div>
</div>
`;
